src/Spectre.libWavelet/WaveletDecomposerRef.cpp

```diff
@@ -0,0 +1,126 @@
+{
+    return 1ULL << (level + 1); // 2^(L + 1)
+}
+
+static constexpr size_t CeiledDivision(size_t dividend, size_t divisor)
+{
+    return dividend / divisor + (dividend % divisor != 0);
+}
+
+static inline size_t ComputeBlockLength(unsigned level, size_t scale, size_t signalLength)
+{
+    constexpr unsigned BASIS_LENGTH = (WAVELET_BASIS * 2) - 1;
+
+    const size_t numerator = signalLength + ((1 << level) - 1) * BASIS_LENGTH;
+    const size_t denominator = scale;
+    const size_t blockLength = CeiledDivision(numerator, denominator) + BASIS_LENGTH;
+
+    return blockLength;
+}
+
+// The function performs an 'intelligent' downsampling, where every second coefficient is
+// stored as a coefficient of 'new' sample, instead of just being discarded. This allows
+// the decomposition to retain more information about the signal.
+static inline void DownsampleByDissolving(CoefficientsPerLevel& coefficients, unsigned level, unsigned blockLength)
+{
+    const unsigned shift = (1 << level) >> 1;
+    for (unsigned i = 0; i < shift; i++)
+    {
+        coefficients[shift + i].resize(blockLength);
+        unsigned oldRowColumn = 0;
+        unsigned newRowColumn = 0;
+        for (unsigned j = 0; j < coefficients[i].size(); j++)
+        {
+            unsigned row = j % 2 ? shift + i : i;
+            unsigned column = j % 2 ? newRowColumn++ : oldRowColumn++;
+            coefficients[row][column] = coefficients[i][j];
+        }
+        coefficients[i].resize(oldRowColumn); // Used to fill moved entries with zeros.
+        coefficients[i].resize(blockLength);
+    }
+}
+
+// Applies the high and low pass filters on the current low-frequency
+// wavelet coefficients, at the level determined by the caller.
+// https://upload.wikimedia.org/wikipedia/commons/1/16/Wavelets_-_SWT_Filter_Bank.png
+// The figure above depicts the process, where h_j prepresents a high- and g_j a low-pass
+// filters, applied at level j.
+// All the low frequency coefficients are overridden on each call, while the high
+// frequnecy ones are just added.
+inline void WaveletDecomposerRef::ApplyFilters(WaveletCoefficients& coefficients,
+    CoefficientsPerLevel& lowFrequencyCoefficients, size_t scale, size_t blockLength, unsigned level) const
+{
+    coefficients.data[level].resize(scale); // Adjust number of coefficient lists for that certain level
+    for (unsigned i = 0; i < scale; i++)
+    {
+        lowFrequencyCoefficients[i].resize(blockLength); // Extend with zeros/shrink to match size
+        coefficients.data[level][i] = // Apply Daubechies4 decomposition high pass filter
+            m_Convolution.Convolve(precomputed::DecompositionHighPassFilter, lowFrequencyCoefficients[i]);
+        lowFrequencyCoefficients[i] = // Apply Daubechies4 decomposition low pass filter
+            m_Convolution.Convolve(precomputed::DecompositionLowPassFilter, lowFrequencyCoefficients[i]);
+    }
+}
+
+WaveletDecomposerRef::WaveletDecomposerRef() : m_Convolution()
```

Comment:
And in such a case I wonder, whether `Convolution` should be a `class` indeed. Class does no partial application. But let's leave it this way now.

Reply:
I just wasn't sure at the beginning. Thought we'll use class to provide some arguments once.